File: Codechef/LUCKY4.cc
#include <cstdio>
#include <ext/hash_map>
#include <cstring>
#include <vector>

using namespace std;
using namespace __gnu_cxx;

struct Tag {
    int n, m, f, prev, limited;

    Tag() {}
    Tag(int n, int m, int f, int p, int l) 
        : n(n), m(m), f(f), prev(p), limited(l) {}

    size_t operator ()(const Tag &t) const {
        return t.n*50
            +  t.m*50*50
            +  t.f*50*50*50
            +  t.prev*50*50*50*50
            +  t.limited*50*50*50*50*50;
    }

    bool operator ==(const Tag &other) const {
        return n == other.n
            && m == other.m
            && f == other.f
            && prev == other.prev
            && limited == other.limited;
    }
};

inline bool lucky(int d) {
    return d == 4 || d == 7;
}

inline int un_f(int f) {
    if(f < 0) return 0;
    return f;
}

int N;
vector<int> limit, c;
hash_map<Tag,long long,Tag> memo;
long long solve(int n, int m, int f, int prev, bool limited) {
    if(n == N) return 1;
    if(m == -1) {
        if(n > 0 && (((f==prev)!=0) != c[n-1])) return 0;
        return solve(n+1, (int)limit.size()-1, -1, f, true);
    }

    const Tag key(n, m, f, prev, limited);
    if(memo.count(key)) return memo[key];

    long long res = 0;
    int lim = limited ? limit[m] : 9;
    int bot = (m==0 && f==-1) ? 1 : 0;
    for(int d = bot; d <= lim; ++d) {
        int nf = lucky(d) ? un_f(f)+1 : (f==-1 && d==0) ? -1 : un_f(f);
        bool nl = limited && (d==lim);
        res += solve(n, m-1, nf, prev, nl);
        if(res > 2000000000LL) break;
    }
    printf("%lld\n", res);
    return memo[key] = res;
}

int main() {
    int T;
    scanf("%d", &T);
    while(T--) {
        int M, K;
        scanf("%d %d %d", &N, &M, &K);
        c.clear();
        for(int i = 0; i < N-1; ++i) {
            int v;
            scanf("%d", &v);
            c.push_back(v==1);
        }
        limit.clear();
        {
            int tmp = M;
            while(tmp) {
                limit.push_back(tmp%10);
                tmp /= 10;
            }
        }
        memo.clear();

        vector<int> ans;
        int prevf = 0;
        long long acc = 0;
        for(int i = 0; i < N; ++i) {
            int num = 0;
            bool limited = true;
            int f = -1;
            for(int dpos = (int)limit.size()-1; dpos >= 0; --dpos) {
                int lim = limited ? limit[dpos] : 9;
                int bot = (dpos==0 && f==-1) ? 1 : 0;
                int digit = -1;
                for(int d = bot; d <= lim; ++d) {
                    int nf = lucky(d) ? un_f(f)+1 : (f==-1 && d==0) ? -1 : un_f(f);
                    bool nl = limited && (d==lim);
                    long long tmp = solve(i, dpos-1, nf, prevf, nl);
                    if(acc+tmp >= K) {
                        digit = d;
                        f = nf;
                        break;
                    }
                    acc += tmp;
                }
                if(digit == -1) {
                    ans.clear();
                    ans.push_back(-1);
                    goto end;
                }

                num *= 10;
                num += digit;
                limited = limited && (digit==lim);
            }
            ans.push_back(num);
            prevf = f;
        }

end:
        bool first = true;
        for(vector<int>::iterator it = ans.begin(); it != ans.end(); ++it) {
            if(!first) printf(" ");
            printf("%d", *it);
            first = false;
        }
        puts("");
    }

    return 0;
}

File: Codechef/LUCKY4.rb
#!/usr/bin/ruby

class Numeric
  def lucky?
    self == 4 || self == 7
  end
end

$maxn = 0
$limit = []
$c = []
$memo = {}

# n番目の数のm桁目以降を自由に決めるときのパターン数．
# fはn番目の数の中でのLucky Number数．
# prevはn-1番目のLucky Number数．
# limitedがtrueのときは，全体でMを越えないようにするためm桁目の数字に上限を設ける．
def solve(n, m, f, prev, limited)
  #puts "#{n} #{m} #{f} #{prev} #{limited}"

  # Base case
  if n == $maxn
    return 1 
  end

  # n番目の数字が決まったら正当性チェックする．
  if m == -1
    return 0 if n > 0 && (f == prev) != $c[n-1]
    return solve(n+1, $limit.size-1, nil, f, true)
  end

  key = [n,m,f,prev,limited]
  res = $memo.fetch(key, nil)
  return res if res

  # 使える数字の上限と下限．
  # A[n] >= 1でないといけないことに注意．
  lim = limited ? $limit[m] : 9
  bot = (m == 0 && f.nil?) ? 1 : 0
  #puts "#{n} #{m} #{f} #{prev} #{limited}: lim=#{lim}"
  
  res = 0
  (bot..lim).each do |d|
    nf = d.lucky? ? f.to_i+1 : (f.nil? && d == 0) ? nil : f.to_i
    nl = limited && (d == lim)
    res += solve(n, m-1, nf, prev, nl)
    # 高速化のための枝刈り．
    # Mは高々10^9なので，それを越えるパターン数を求めても仕方がない．
    # （単純に再帰が増えるだけではなく，Bignumになってしまうと激遅になる）
    break if res > 2000000000
  end
  #puts "#{n} #{m} #{f} #{prev} #{limited}: #{res}"
  $memo[key] = res
end

gets.to_i.times do
  $maxn, m, k = gets.split.map(&:to_i)
  $c = gets.split.map{|c| c == "1"}
  $limit = m.to_s.each_char.to_a.map(&:to_i).reverse
  $memo.clear

  ans = []
  prevf = 0
  acc = 0
  catch(:end) do
    $maxn.times do |pos|
      num = 0
      limited = true
      f = nil
      # 上の桁から固定していき，それ以降のパターン数を数える．
      (0..$limit.size-1).to_a.reverse_each do |dpos|
        lim = limited ? $limit[dpos] : 9
        bot = (dpos == 0 && f.nil?) ? 1 : 0
        digit = nil
        (bot..lim).each do |d|
          nf = d.lucky? ? f.to_i+1 : (f.nil? && d == 0) ? nil : f.to_i
          nl = limited && (d==lim)
          tmp = solve(pos, dpos-1, nf, prevf, nl)
          #puts "#{d}: #{acc+tmp}"
          if acc + tmp >= k
            digit = d
            f = nf
            break
          end
          acc += tmp
        end
        if digit.nil?
          ans = [-1]
          throw :end
        end

        num *= 10
        num += digit
        limited = limited && (digit == lim)
      end
      ans << num
      prevf = f
    end
  end

  puts ans.join(" ")
end
